### DISCUSSION TITLE
Process outbox messages on a server cluster

### DISCUSSION THREAD

**Discussion by patxiba** [kgrzybek/modular-monolith-with-ddd#258] [General]
What is the solution for the outbox message processor not to repeat events if the API is installed on a server cluster.
In this scenario, Quarzt runs two instances on different machines, so each instance can run the same event.

**Comment by LukasKuchta** [comment]
Good question! How to solve this issue when scale up some module horizontaly?

**Reply by senal** [reply]
All consumers must be idempotent.  means, it should be able to handle duplicate requests.
hope the below articles will help you to grasp the idea.

https://serverlessland.com/event-driven-architecture/idempotency

https://levelup.gitconnected.com/idempotency-in-restful-apis-and-event-driven-systems-43a26f91ea8d

Implementation of this pattern may vary on the tech stack and the architectures at your disposal.

Cheers

**Reply by LukasKuchta** [reply]
@senal thanks this is exactly what i want.

What are the possible approaches for "locking" when i read/write un-proccessed messages for multiple instances?  I can use transaction for read messages and then write handling state. But other instances will wait until first release the transaction. When operation inside the transaction is long, this is not best solution. Becouse other instances can perform actions parallel way. Is there any others approaches ?

**Reply by Minh8181-blz** [reply]
I have the same concern. Besides, even if we already implement idempotency, what is the solution to minimize the number of duplicate fetching with Quartz?
